health/http.go

```diff
@@ -0,0 +1,108 @@
+package health
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	defaultHealthCheckRoute    = "/health"
+	defaultReadinessCheckRoute = "/ready"
+	defaultPort                = 4444
+)
+
+type CheckFunc func() error
+
+type Option func(*server)
+
+type server struct {
+	healthCheckRoute    string
+	readinessCheckRoute string
+	port                int
+	healthCheckFunc     CheckFunc
+	readinessCheckFunc  CheckFunc
+}
+
+func WithHealthCheckRoute(route string) Option {
+	return func(s *server) {
+		s.healthCheckRoute = route
+	}
+}
+
+func WithReadinessCheckRoute(route string) Option {
+	return func(s *server) {
+		s.readinessCheckRoute = route
+	}
+}
+
+func WithPort(port int) Option {
+	return func(s *server) {
+		s.port = port
+	}
+}
+
+func WithHealthCheckFunc(healthCheckFunc CheckFunc) Option {
+	return func(s *server) {
+		s.healthCheckFunc = healthCheckFunc
+	}
+}
+
+func WithReadinessCheckFunc(readinessCheckFunc CheckFunc) Option {
+	return func(s *server) {
+		s.readinessCheckFunc = readinessCheckFunc
+	}
+}
+
+func handle(handler *http.ServeMux, route string, handleFunc CheckFunc) {
+	handler.HandleFunc(route, func(w http.ResponseWriter, r *http.Request) {
+		if handleFunc == nil {
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+
+		if err := handleFunc(); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	})
+}
+
+// StartHealthCheckServer starts a HTTP server to handle health check and readiness check requests.
+func StartHealthCheckServer(ctx context.Context, opts ...Option) error {
+	hcServer := &server{
+		healthCheckRoute:    defaultHealthCheckRoute,
+		readinessCheckRoute: defaultReadinessCheckRoute,
+		port:                defaultPort,
+	}
+
+	for _, opt := range opts {
+		opt(hcServer)
+	}
+
+	handler := http.NewServeMux()
+	handle(handler, hcServer.healthCheckRoute, hcServer.healthCheckFunc)
+	handle(handler, hcServer.readinessCheckRoute, hcServer.readinessCheckFunc)
+
+	srv := &http.Server{
+		Addr:    fmt.Sprintf(":%d", hcServer.port),
+		Handler: handler,
+	}
+
+	go func() {
+		<-ctx.Done()
+		if err := srv.Shutdown(ctx); err != nil {
+			log.Info("server shutdown: ", err)
+		}
+	}()
+
+	if err := srv.ListenAndServe(); err != http.ErrServerClosed {
```

Comment:
what are the pros of running healthcheck handlers on a different port as a standalone server, rather than adding another endpoint/handler to the main http server? 
Are we trying to cover cases where an app doesn't run a http server by itself, so we run it for them?

Also if we run healthcheck endpoint on a different port, in the passed checkers clients should test the main http server port as well. I think it makes sense to document this explicitly.

Reply:
another case - what if a single binary runs multiple http servers that have their own dependencies. In this case a single healthcheck server won't work well. Thus clients will have to setup one healthcheck server for each running "app" http server.

Reply:
> what are the pros of running healthcheck handlers on a different port as a standalone server, rather than adding another endpoint/handler to the main http server? Are we trying to cover cases where an app doesn't run a http server by itself, so we run it for them?
> 

yes, this lib allows to add health checks to apps like consumers or generic workers.

> Also if we run healthcheck endpoint on a different port, in the passed checkers clients should test the main http server port as well. I think it makes sense to document this explicitly.

this lib is meant to be generic and can be used in different apps, so health check implementation really depends on the app